client/src/controllers/PreviewController.ts

```diff
@@ -0,0 +1,473 @@
+   * @param event The click event
+   * @returns whether the data is valid
+   */
+  async handlePreviewInNewTab(event: MouseEvent) {
+    event.preventDefault();
+    const link = event.currentTarget as HTMLAnchorElement;
+
+    const valid = await this.handlePreview();
+    if (valid) {
+      // Use the base URL value (without any params) as the target (identifier)
+      // for the window, so that if the user switches between preview modes,
+      // the same window will be reused.
+      window.open(new URL(link.href).toString(), this.baseUrlValue) as Window;
+    }
+    return valid;
+  }
+
+  /**
+   * Checks whether the form data has changed since the last call to this method
+   * @returns whether the form data has changed
+   */
+  hasChanges() {
+    // https://github.com/microsoft/TypeScript/issues/30584
+    const newPayload = new URLSearchParams(
+      new FormData(this.editForm) as unknown as Record<string, string>,
+    ).toString();
+    const changed = this.formPayload !== newPayload;
+
+    this.formPayload = newPayload;
+    return changed;
+  }
+
+  /**
+   * Handle the side panel show event. The preview data is immediately updated.
+   * Then, if auto update is enabled, apply debounce to setPreviewData for
+   * subsequent calls, and set up an interval to automatically check the form
+   * and update the preview data as necessary.
+   */
+  handlePanelShown() {
+    // Immediately update the preview when the panel is opened
+    this.checkAndUpdatePreview();
+
+    // Skip setting up the interval if auto update is disabled
+    if (!this.autoUpdateValue) return;
+
+    // Apply debounce for subsequent updates if not already applied
+    if (!('cancel' in this.setPreviewData)) {
+      this.setPreviewData = debounce(
+        this.setPreviewData.bind(this),
+        this.autoUpdateIntervalValue,
+      );
+    }
+
+    // Only set the interval while the panel is shown
+    // This interval performs the checks for changes but not necessarily the
+    // update itself
+    this.updateInterval = setInterval(
+      this.checkAndUpdatePreview.bind(this),
+      this.autoUpdateIntervalValue,
+    );
+  }
+
+  /**
+   * Handle the side panel hide event. Clear the auto-update interval (if any).
+   */
+  handlePanelHidden() {
+    if (!this.updateInterval) return;
+    clearInterval(this.updateInterval);
+    this.updateInterval = null;
+  }
+
+  /**
+   * Changes the preview mode specified in the iframe and new tab link,
+   * and updates the preview.
+   * @param event Event from the `<select>` element
+   */
+  handlePreviewModeChange(event: Event) {
+    const mode = (event.target as HTMLSelectElement).value;
+    const url = new URL(this.baseUrlValue, window.location.href);
+
+    // Update the new tab link
+    url.searchParams.set('mode', mode);
+    this.newTabButtonTarget.href = url.toString();
+
+    // Make sure data is updated and an alert is displayed if an error occurs
+    this.handlePreview();
+  }
+
+  connect() {
```

Comment:
Not strict but it's nice to have the connect method at the top (after declaring properties), as it's kind of the entry point to the controller.